### PR TITLE
Add helm chart for APIM 3.1.0 pattern 1 and 2

### PR DESCRIPTION
**Pull request by Chathuru**
## Description
Create Kubernetes resources for WSO2 API Manager version 3.1.0. These resources can be used to deploy APIM deployment patterns 1 and 2. Resolving the issue https://github.com/wso2/kubernetes-apim/issues/330

[1] - https://apim.docs.wso2.com/en/latest/install-and-setup/deploying-wso2-api-manager/deployment-patterns/#wso2-api-manager-deployment-patterns

### CODE REVIEW

**Comment by chirangaalwis** [comment]
@Chathuru please integrate the changes from PRs https://github.com/wso2/kubernetes-apim/pull/333 and https://github.com/wso2/kubernetes-apim/pull/336 into these resources.

File: advanced/am-pattern-1/requirements.yaml [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# Copyright (c) 2019, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+#Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (chirangaalwis) [inline advanced/am-pattern-1/requirements.yaml]: @Chathuru please adjust according to the previously mentioned license header setting rule.

File: advanced/am-pattern-2/templates/wso2am-pattern-2-secrets.yaml [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+apiVersion: v1
+kind: Secret
+metadata:
+  name: wso2am-pattern-2-wso2-private-registry-creds
+  namespace: {{ .Release.Namespace }}
+type: kubernetes.io/dockerconfigjson
+data:
+  .dockerconfigjson: {{ $files.Get "auth.json" | replace "reg.id" $regId | replace "reg.username" $username | replace "reg.password" $password | replace "reg.email" $email | replace "reg.auth" $auth | b64enc }}
+
+  {{ end }}
```
Comment (chirangaalwis) [inline advanced/am-pattern-2/templates/wso2am-pattern-2-secrets.yaml]: @Chathuru please add a new line at EOF.

File: advanced/am-pattern-2/values.yaml [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+            maxSurge: 2
+            # The maximum number of pods that can be unavailable during the update.
+            maxUnavailable: 0
+#        # If the deployment configurations for the Gateway profile of WSO2 API Manager v3.0.0 (<WSO2AM>/repository/conf/deployment.toml),
+#        # add the customized configuration file under (wso2 -> deployment -> am -> gateway -> config -> deployment.toml)
+#        config:
+#          deployment.toml: |-
+#            # deployment configurations for the Gateway profile of WSO2 API Manager v3.0.0 (<WSO2AM>/repository/conf/deployment.toml)
+      # API Manager's Key Manager specific configurations
+      km:
```
Comment (chirangaalwis) [inline advanced/am-pattern-2/values.yaml]: @Chathuru did you integrate support for switching between API Manager's Key Manager and IS as Key Manager, as we had done in previous versions of Kubernetes resources.

@pubudu538 do we need to integrate support for both or do we now have a recommended KM for this type of deployments?

Comment (pubudu538) [reply advanced/am-pattern-2/values.yaml]: Better to include both.

Comment (chirangaalwis) [reply advanced/am-pattern-2/values.yaml]: @pubudu538 ack.

@Chathuru please note.

File: advanced/am-pattern-1/Chart.yaml [hunk 4]
```diff
@@ -13,7 +13,13 @@
...
 
 apiVersion: v1
-appVersion: "3.0.0"
+appVersion: "3.1.0"
 description: A Helm chart for the deployment of WSO2 APIM-Analytics deployment
 name: am-pattern-1
-version: 3.0.0-3
+version: 3.1.0
+
+#dependencies:
```
[the remark this answers is not in the document]
Comment (Chathuru) [reply advanced/am-pattern-1/Chart.yaml]: Uncommented the dependency https://github.com/Chathuru/kubernetes-apim/commit/1d4d91b99069f609dea7e266cce59f9fa431b59d commit. However, we need to upload the mysql-am version 3.0.0-4

File: advanced/am-pattern-1/Chart.yaml [hunk 5]
```diff
@@ -1,4 +1,4 @@
-# Copyright (c) 2018, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+# Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (Chathuru) [reply advanced/am-pattern-1/Chart.yaml]: Fix the versioning on https://github.com/Chathuru/kubernetes-apim/commit/1d4d91b99069f609dea7e266cce59f9fa431b59d commit

File: README.md [hunk 6]
```diff
@@ -21,7 +22,7 @@
 
 ## Changelog
 
-**Change log** from previous v3.0.0.2 release: [View Here](CHANGELOG.md)
+**Change log** from previous v3.1.0.1 release: [View Here](CHANGELOG.md)
```
[the remark this answers is not in the document]
Comment (Chathuru) [reply README.md]: Fix the versioning on https://github.com/Chathuru/kubernetes-apim/commit/1d4d91b99069f609dea7e266cce59f9fa431b59d commit

File: advanced/am-pattern-1/README.md [hunk 7]
```diff
@@ -76,16 +76,16 @@ If you do not have an active WSO2 subscription, do not change the parameters `ws
...
+| `wso2.deployment.am.resources.requests.memory`                              | The minimum amount of memory that should be allocated for a Pod                           | 4Gi                         |
 | `wso2.deployment.am.resources.requests.cpu`                                 | The minimum amount of CPU that should be allocated for a Pod                              | 2000m                       |
-| `wso2.deployment.am.resources.limits.memory`                                | The maximum amount of memory that should be allocated for a Pod                           | 3Gi                         |
-| `wso2.deployment.am.resources.limits.cpu`                                   | The maximum amount of CPU that should be allocated for a Pod                              | 3000m                       |
+| `wso2.deployment.am.resources.limits.memory`                                | The maximum amount of memory that should be allocated for a Pod                           | 4Gi                         |
```
[the remark this answers is not in the document]
Comment (Chathuru) [reply advanced/am-pattern-1/README.md]: Revert the requirement in Fix the versioning on https://github.com/Chathuru/kubernetes-apim/commit/1d4d91b99069f609dea7e266cce59f9fa431b59d commit commit

File: advanced/am-pattern-1/confs/am-analytics-dashboard/bin/carbon.sh [hunk 8]
```diff
@@ -1,6 +1,6 @@
 #!/bin/sh
 # ---------------------------------------------------------------------------
-#  Copyright (c) 2015, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+#  Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (Chathuru) [reply advanced/am-pattern-1/confs/am-analytics-dashboard/bin/carbon.sh]: Fix the versioning on https://github.com/Chathuru/kubernetes-apim/commit/1d4d91b99069f609dea7e266cce59f9fa431b59d commit

File: advanced/am-pattern-1/requirements.yaml (hunk 1 above)
Comment (Chathuru) [reply advanced/am-pattern-1/requirements.yaml]: Fix the versioning on https://github.com/Chathuru/kubernetes-apim/commit/1d4d91b99069f609dea7e266cce59f9fa431b59d commit

File: advanced/am-pattern-2/values.yaml (hunk 3 above)
Comment (Chathuru) [reply advanced/am-pattern-2/values.yaml]: On this https://github.com/wso2/kubernetes-apim/pull/332 I only add API Manager's Key Manager. Below changes will add in a separately,
1. Add IS-as-KM
2. Add StatefulSet
3. Add Analytics HA deployment

File: advanced/am-pattern-1/README.md [hunk 9]
```diff
@@ -1,4 +1,4 @@
-# Helm Chart for deployment of WSO2 API Manager with WSO2 API Manager Analytics
+# Helm Chart for deployment of WSO2 API Manager Pattern 1: Single node (all-in-one) deployment
```
Comment (chirangaalwis) [inline advanced/am-pattern-1/README.md]: @Chathuru please include details about WSO2 product Docker image types in the README documentation. Please restructure the README as per [1]. Feel free to drop in your suggestions and concerns.

[1]: [IAM pattern 1 README documentation](https://github.com/wso2/kubernetes-is/blob/v5.10.0.1/advanced/is-pattern-1/README.md)

File: advanced/am-pattern-2/README.md [hunk 10]
```diff
@@ -0,0 +1,203 @@
+# Helm Chart for deployment of WSO2 API Manager Pattern 2: Deployment with a separate Gateway and separate Key Manager
```
Comment (chirangaalwis) [inline advanced/am-pattern-2/README.md]: @Chathuru please include details about WSO2 product Docker image types in the README documentation. Please restructure the README as per [1]. Feel free to drop in your suggestions and concerns.

[1]: [IAM pattern 1 README documentation](https://github.com/wso2/kubernetes-is/blob/v5.10.0.1/advanced/is-pattern-1/README.md)

File: advanced/am-pattern-2/templates/wso2am-pattern-2-secrets.yaml (hunk 2 above)
Comment (chirangaalwis) [reply advanced/am-pattern-2/templates/wso2am-pattern-2-secrets.yaml]: @Chathuru you have missed this.

File: advanced/am-pattern-2/values.yaml (hunk 3 above)
Comment (chirangaalwis) [reply advanced/am-pattern-2/values.yaml]: @Chathuru 2 can be integrated in a new release.

But improvements 1 and 3 ought to be released as part of the initial release, IMO.

**Pull request closed without merging**